### PR TITLE
fix torch.nn.parallel.scatter_gather.gather to handle NamedTuples and handle moving output to CPU

### PR DESCRIPTION
**Pull request by ajsanjoaquin**
Fixes #{[50510](https://github.com/pytorch/pytorch/issues/50510)}

Allows ```torch.nn.parallel.scatter_gather.gather``` to accept a list of NamedTuples as input and returns a NamedTuple whose elements are tensors. I added the author's fix using the ```is_namedtuple``` function. 

While testing this fix, I encountered a deprecation warning instructing me to use ```'cpu'``` instead of ```-1``` to move the outputs to the CPU. However, doing this causes an assertion error in the ```_get_device_index``` function. I solved this by handling the CPU case in the affected ```forward``` function.
@rohan-varma

### CODE REVIEW

File: test/test_cuda.py [hunk 1]
```diff
@@ -3496,6 +3496,50 @@ class TestNamedTupleInput_1(NamedTuple):
...
+        a = torch.rand(num_gpus * 2, device=0)
+        b = torch.rand(num_gpus * 2, device=1)
+        out1 = TestNamedTupleInput_0(a, b)
+
+        a = torch.rand(num_gpus * 2, device=1)
+        b = torch.rand(num_gpus * 2, device=0)
+        out2 = TestNamedTupleInput_0(a, b)
+
+        outputs = [out1, out2]
+        out = scatter_gather.gather(outputs, 'cpu')  # test on cpu
```
Comment (rohan-varma) [inline test/test_cuda.py:3516]: Should we have a test for GPU here, and a test for CPU for the one below so all paths are exercised?

File: torch/nn/parallel/_functions.py [hunk 2]
```diff
@@ -56,8 +56,12 @@ def forward(ctx, target_device, dim, *inputs):
         assert all(i.device.type != 'cpu' for i in inputs), (
             'Gather function not implemented for CPU tensors'
         )
-        target_device = _get_device_index(target_device, True)
-        ctx.target_device = target_device
+        if (target_device == 'cpu'):
+            target_device = _get_device_index(target_device, True, allow_cpu=True)
+            ctx.target_device = 'cpu'
```
Comment (rohan-varma) [inline torch/nn/parallel/_functions.py:61]: Shouldn't this be `ctx.target_device = target_device`? And in general can't the change be reverted to what it already was? 

If not, then is there a need to modify `target_device` for the CPU path?

**Review by rohan-varma — COMMENTED** [review]
Looks good overall, just had a few questions inline. Thanks for working on this!

File: torch/nn/parallel/_functions.py (hunk 2 above)
Comment (ajsanjoaquin) [reply torch/nn/parallel/_functions.py:61]: Setting ```ctx.target_device = target_device``` sets ```ctx.target_device ``` to ```-1```, which triggers the Deprecation warning in ```comm.gather```. I'm not sure how to set ```target_device = 'cpu'``` in the cpu case , so I explicitly set the ```ctx.target_device``` to cpu.

File: test/test_cuda.py (hunk 1 above)
Comment (ajsanjoaquin) [reply test/test_cuda.py:3516]: I'm not sure what you mean. Do you mean have either ```a``` or ```b``` tensor in the CPU and test if the gather function can gather from those devices?

Comment (rohan-varma) [reply test/test_cuda.py:3516]: I meant that we should always have a test as follows:
`out = scatter_gather.gather(outputs, 0)`

basically similar to the test on line 3536, but for the `TestnamedtupleInput_0` type. And vice versa, the below test should also have `out = scatter_gather.gather(outputs, 'cpu')`?

Comment (ajsanjoaquin) [reply test/test_cuda.py]: Ah I see. Those tests would ensure that a previous call to ```scatter_gather.gather``` does not affect the output of future calls to the function. I have added those tests!

File: torch/nn/parallel/_functions.py (hunk 2 above)
Comment (ajsanjoaquin) [reply torch/nn/parallel/_functions.py:60]: Removed line 60 because it's unnecessary

**Comment by ajsanjoaquin** [comment]
Some tests are failing: 3 because of merge-conflicts in folders I did not modify, and 1 because of some unit tests that I did not modify, nor is affected by the modified functions. Is this something that should be fixed on my end?

File: torch/nn/parallel/scatter_gather.py [hunk 3]
```diff
@@ -54,8 +54,8 @@ def scatter_kwargs(inputs, kwargs, target_gpus, dim=0):
 
 def gather(outputs, target_device, dim=0):
     r"""
-    Gathers tensors from different GPUs on a specified device
-      (-1 means the CPU).
+    Gathers tensors from different GPUs on a specified device.
+    (Use 'cpu' for CPU)
```
Comment (rohan-varma) [inline torch/nn/parallel/scatter_gather.py:58]: Nit: Might be good to add a comment about the deprecation warning you encountered to explain this for future readers

**Review by rohan-varma — APPROVED** [review]

**Comment by rohan-varma** [comment]
> Some tests are failing: 3 because of merge-conflicts in folders I did not modify, and 1 because of some unit tests that I did not modify, nor is affected by the modified functions. Is this something that should be fixed on my end?

It looks like the CI is flaky, I don't think it's related to this diff. Although, you mind rebasing/merging master to retrigger CI just to make sure? Thanks!

**Comment by rohan-varma** [comment]
Looks good! Lint errors seem unrelated - 

```
[
  {
    path: 'test/test_autograd.py',
    start_line: 33,
    end_line: 33,
    start_column: 1,
    end_column: 1,
    annotation_level: 'failure',
    message: "[F811] redefinition of unused 'gradcheck' from line 25"
  },
  {
    path: 'test/test_autograd.py',
    start_line: 33,
    end_line: 33,
    start_column: 1,
    end_column: 1,
    annotation_level: 'failure',
    message: "[F811] redefinition of unused 'gradgradcheck' from line 25"
  }
]
```

**Pull request closed without merging**